### ISSUE TITLE
cache#increment fails if key exists and has been pre-configured to 0 

### ISSUE THREAD

**Issue by stevegoobermanhill** [redis-store/redis-store#253]
Hi,
I'm running Rails 4.1.6 on Ruby 2.2.1p85 with Redis 3.0.3

It looks like the underlying Redis is failing and throwing an error if I try and increment a counter in the store that has been pre-initialized to 0.

[1] pry(main)> Rails.cache.increment('foo')
=> 1
[2] pry(main)> Rails.cache.write('bar',0)
=> "OK"
[3] pry(main)> Rails.cache.increment('bar')
Redis::CommandError: ERR value is not an integer or out of range
from /home/admin2/.rvm/gems/ruby-2.2.1/gems/redis-3.2.1/lib/redis/client.rb:113:in `call'
[4] pry(main)> 

Is this a known behaviour ?

Kind regards

Steve

**Comment by krainboltgreene** [comment]
Interesting. I'll see if I can replicate soon.

**Comment by hmcfletch** [comment]
TL;DR; - Read and write using `raw: true` for things you are going to increment.
```
> Rails.cache.write('foo', 0, raw: true)
> Rails.cache.read('foo', raw: true)
> Rails.cache.increment('foo', 3)
```


I ran into a similar issue. Running these commands helped me figure it out:

```
> Rails.cache.read('bar')
=> nil
> Rails.cache.write('bar', 0)
=> "OK"
> Rails.cache.read('bar')
=> 0
> Rails.cache.read('bar', raw: true)
=> "\u0004\bo: ActiveSupport::Cache::Entry\b:\v@valuei\u0000:\u0010@created_atf\u00171454098794.0263748:\u0010@expires_inf\n5.4e3"
> Rails.cache.increment('bar', 1)
Redis::CommandError: ERR value is not an integer or out of range
   ...
```

Looks like the entry is getting wrapped and marshaled before writing to the Redis, so when you try to do a native increment, it is trying to increment a String representing a marshalled ruby object.

Another interesting set of commands:

```
> Rails.cache.read('foo')
=> nil
> Rails.cache.write('foo', 0, raw: true)
=> "OK"
> Rails.cache.read('foo')
ArgumentError: marshal data too short
   ...
> Rails.cache.read('foo', raw: true)
=> "0"
> Rails.cache.increment('foo', 1)
=> 1
> Rails.cache.read('foo', raw: true)
=> "1"
```

If you write with `raw: true` you have to read with `raw: true` as well, otherwise it try to unmarshal and fails.

**Comment by hmcfletch** [comment]
Looks like #231 has something to do with it as well

**Comment by tubbo** [comment]
i can take this on...will try and get that other PR cleaned up & merged.

**Comment by d4rky-pl** [comment]
```
> Rails.cache.increment("foo", 1)
=> 1
> Rails.cache.read("foo")
ArgumentError: marshal data too short
```
